common/src/lru.rs

```diff
@@ -1,925 +1,286 @@
+            }
 
-        Some((key, val))
-    }
-}
+            let entry_index = self.cache.used_indexes[self.pos];
+            let entry = &self.cache.entries[entry_index];
 
-impl<'a, K, V> ExactSizeIterator for Iter<'a, K, V> {}
-impl<'a, K, V> FusedIterator for Iter<'a, K, V> {}
+            self.pos += 1;
 
-impl<'a, K, V> Clone for Iter<'a, K, V> {
-    fn clone(&self) -> Iter<'a, K, V> {
-        Iter {
-            len: self.len,
-            ptr: self.ptr,
-            end: self.end,
-            phantom: PhantomData,
+            if let Some((ref k, ref v)) = entry {
+                return Some((&k, &v));
+            }
         }
     }
 }
 
-// The compiler does not automatically derive Send and Sync for Iter because it contains
-// raw pointers.
-unsafe impl<'a, K: Send, V: Send> Send for Iter<'a, K, V> {}
-unsafe impl<'a, K: Sync, V: Sync> Sync for Iter<'a, K, V> {}
-
-/// An iterator over mutables entries of a `LruCache`.
-///
-/// This `struct` is created by the [`iter_mut`] method on [`LruCache`][`LruCache`]. See its
-/// documentation for more.
-///
-/// [`iter_mut`]: struct.LruCache.html#method.iter_mut
-/// [`LruCache`]: struct.LruCache.html
-pub struct IterMut<'a, K: 'a, V: 'a> {
-    len: usize,
-
-    ptr: *mut LruEntry<K, V>,
-    end: *mut LruEntry<K, V>,
-
-    phantom: PhantomData<&'a K>,
+/// Mutable Iterator over values in an LruCache, from most-recently-used to least-recently-used.
+pub struct LruCacheMutIterator<'a, K: 'a, V: 'a> {
+    cache: &'a mut LruCache<K, V>,
+    pos: usize,
 }
 
-impl<'a, K, V> Iterator for IterMut<'a, K, V> {
+impl<'a, K, V> Iterator for LruCacheMutIterator<'a, K, V> {
     type Item = (&'a K, &'a mut V);
 
-    fn next(&mut self) -> Option<(&'a K, &'a mut V)> {
-        if self.len == 0 {
-            return None;
-        }
-
-        let key = unsafe { &(*self.ptr).key };
-        let val = unsafe { &mut (*self.ptr).val };
+    fn next(&mut self) -> Option<Self::Item> {
+        loop {
+            if self.pos == self.cache.used_indexes.len() {
+                return None;
+            }
 
-        self.len -= 1;
-        self.ptr = unsafe { (*self.ptr).next };
+            let entry_index = self.cache.used_indexes[self.pos];
 
-        Some((key, val))
-    }
+            // Hack to get around the borrow checker. The borrow checker would prevent us from
+            // returning a mutable reference to an entry, since in order to get the entry we need
+            // to hold a reference to the Option inside `self.cache.entries`. However, since we're
+            // holding that reference, we will then be denied from creating another mutable
+            // reference.
+            // This code is safe because the list of entries is not going to be modified during
+            // iteration, since a mutable reference to the LruCache is held by the iterator.
+            let entry =
+                unsafe { &mut *(&mut self.cache.entries[entry_index] as *mut Option<(Arc<K>, V)>) };
```

Comment:
I wonder if there is another way around this. I don't have a suggestion at the moment, but I think it would be great to add `#[deny(unsafe)]` to this crate and not have this caveat, even though the comment explaining why it's safe seems sound.

Reply:
I agree that it would be best to avoid `unsafe`, but I also don't have an alternative to offer.

Reply:
I couldn't think of a better way to do it. There's no clean way of holding two mutable references to the same object, it's kind of the point of the borrow checker :)
This is a special occasion where it is guaranteed to not cause a mess, but I am not aware of a non-unsafe way to accomplish this.
There might be some questionable way of using Cell/RefCell to achieve this, but I don't think that is in any way cleaner or better.